_posts/2020-11-2-kubernetes-webhooks-made-easy-with-openfaas-part-2.md

```diff
@@ -0,0 +1,291 @@
+---
+title: "Kubernetes Webhooks made easy with OpenFaaS - Part 2"
+description: "In this post you'll learn how to write Kubernetes Admission webhooks using OpenFaaS functions - Part 2"
+date: 2020-11-2
+image: /images/2020-10-27-k8s-validatingwebhook-openfaas/puzzle.jpg
+categories:
+  - arkade
+  - kubectl
+  - faas-cli
+  - admissionwebhooks
+  - mutatingadmissionwebhooks
+  - k8s extensibility
+author_staff_member: developer-guy
+dark_background: true
+---
+
+In this post you'll learn how to write Kubernetes Admission webhooks using OpenFaaS functions - Part 2
+
+<p align="center">
+<img height="128" src="/images/openfaas/kubernetes.png">
+</p>
+
+## Introduction to Kubernetes Mutating Admission webhooks
+In the previous [part](https://www.openfaas.com/blog/kubernetes-webhooks-made-easy-with-openfaas/), we talked about what Kubernetes Admission Webhooks are and how can we use Validating Admission Webhook.
+
+Today we are gonna talk about Mutating Admission Webhook and demonstrate how can we use it as OpenFaaS Function.
+
+The main difference between the two types of admission webhook are pretty self-explanatory: validating webhooks can reject a request, but they cannot modify the object they are receiving in the admission request, while mutating webhooks can modify objects by creating a patch that will be sent back in the admission response and can reject a request too. If a webhook rejects a request, an error is returned to the end-user.
+
+Also, the other difference between two types are validating webhooks are called in parallel and mutating webhooks are called in-sequence by the api-server.
+
+## The Scenario
+This time we will inject a file to the container filesystem using [configMap volume](https://kubernetes.io/docs/concepts/storage/volumes/#configmap) automatically by the Mutating Admission Webhook.
+
+In order to do that, we need to create a file that will injected by the Mutating Admission Webhook to the container's filesystem like below:
+
+```sh
+$ cat functions/fileinjector/hello-openfaas.txt
```

Comment:
I don't think this specific example is convincing, it seems to be trying too hard to be a solution to a problem that doesn't exist.

Please come up with a plausible, example and use that instead.

My suggestions would be:

* sets a default memory limit for a deployment if one is not present or given in the spec
* what about adding a default rate-limit to all ingress-nginx ingress records? https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/

Both of these are real-world examples that will make sense to the reader.

Reply:
We changed the demo as you said, we are going to set the default resources if no resources found for the deployment by the mutating webhook. 👍

Reply:
Done